### ISSUE TITLE
Trainer.scala

### ISSUE THREAD

**Issue by robinchampseix** [robinchampseix/Spark#1]
package com.sparkProject

import org.apache.spark.SparkConf
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.LogisticRegression
import org.apache.spark.ml.evaluation.{MulticlassClassificationEvaluator}
import org.apache.spark.ml.feature._
import org.apache.spark.ml.tuning.{ParamGridBuilder, TrainValidationSplit}
import org.apache.spark.sql.{DataFrame, SparkSession}

object Trainer {

  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setAll(Map(
      "spark.scheduler.mode" -> "FIFO",
      "spark.speculation" -> "false",
      "spark.reducer.maxSizeInFlight" -> "48m",
      "spark.serializer" -> "org.apache.spark.serializer.KryoSerializer",
      "spark.kryoserializer.buffer.max" -> "1g",
      "spark.shuffle.file.buffer" -> "32k",
      "spark.default.parallelism" -> "12",
      "spark.sql.shuffle.partitions" -> "12",
      "spark.driver.maxResultSize" -> "2g"
    ))

    val spark = SparkSession
      .builder
      .config(conf)
      .appName("TP_spark")
      .getOrCreate()

    /*******************************************************************************
      *
      *       TP 4-5
      *
      *       - lire le fichier sauvegarder précédemment
      *       - construire les Stages du pipeline, puis les assembler
      *       - trouver les meilleurs hyperparamètres pour l'entraînement du pipeline avec une grid-search
      *       - Sauvegarder le pipeline entraîné
      *
      *       if problems with unimported modules => sbt plugins update
      *
      ********************************************************************************/

   /** CHARGER LE DATASET **/
     // On charge les données, préalablement préparées
   val df: DataFrame = spark
     .read
     .parquet("/Users/robinchampseix/Desktop/Telecom/MS_Big_Data/P1/Hadoop_Spark/TP/TP_ParisTech_2017_2018_starter/src/main/scala/com/sparkProject/prepared_trainingset")

    /** TF-IDF **/
    // On affiche les 5 premières lignes de notre dataframe
    df.show(5)
    // Le schema donne le nom et type (string, integer,...) de chaque colonne
    df.printSchema()

    // a) On sépare les textes en mots (ou tokens) avec un tokenizer
    val tokenizer = new RegexTokenizer()
      .setPattern("\\W+")
      .setGaps(true)
      .setInputCol("text")
      .setOutputCol("tokens")

    // b) On retire les stop words pour ne pas encombre le modèle avec des mots qui ne véhiculent pas de sens
    val remover = new StopWordsRemover()
      .setInputCol("tokens")
      .setOutputCol("remover")

    // c) Partie TF : convert a collection of text documents to vectors of token counts
    val cvm = new CountVectorizer()
      .setInputCol("remover")
      .setOutputCol("cvm")

    // d) Partie IDF
    val idf = new IDF()
        .setInputCol("cvm")
        .setOutputCol("tfidf")

    /** Convertir les catégories en données numériques **/
    // e) On convertit la variable catégorielle “country2” en données numérique
    val indexer1b = new StringIndexer()
      .setInputCol("country2")
      .setOutputCol("country2_indexed")
    // This allows algorithms which expect continuous features to use categorical features
    val indexer = new OneHotEncoder()
      .setInputCol("country2_indexed")
      .setOutputCol("country_indexed")

    // f) On convertit la variable catégorielle “currency2” en données numérique
    val indexer2b = new StringIndexer()
      .setInputCol("currency2")
      .setOutputCol("currency2_indexed")
    // This allows algorithms which expect continuous features to use categorical features
    val indexer2 = new OneHotEncoder()
      .setInputCol("currency2_indexed")
      .setOutputCol("currency_indexed")

    /** VECTOR ASSEMBLER **/
    // g) On assemble les features "tfidf", "days_campaign", "hours_prepa", "goal",
    // "country_indexed", "currency_indexed" dans une seule colonne “features”.
    val assembler = new VectorAssembler()
      .setInputCols(Array("tfidf", "days_campaign", "hours_prepa", "goal", "country_indexed", "currency_indexed"))
      .setOutputCol("features")

    /** MODEL **/
    // h) Régression logistique
    val lr = new LogisticRegression()
      .setElasticNetParam(0.0)
      .setFitIntercept(true)
      .setFeaturesCol("features")
      .setLabelCol("final_status")
      .setStandardization(true)
      .setPredictionCol("predictions")
      .setRawPredictionCol("raw_predictions")
      .setThresholds(Array(0.7, 0.3))
      .setTol(1.0e-6)
      .setMaxIter(300)

    /** PIPELINE **/
    // i) Création du pipeline en combinant les 8 étpaes précédentes
    val pipeline = new Pipeline()
      .setStages(Array(tokenizer, remover, cvm, idf,indexer1b,indexer,indexer2b,indexer2,assembler,lr))

    /** TRAINING AND GRID-SEARCH **/

    // j) Création du Training et du Test set
    val Array(training, test) = df.randomSplit(Array(0.9, 0.1))

    // k) GridSearch paramétrée selon les données de l'énoncé
    val paramGrid = new ParamGridBuilder()
      .addGrid(cvm.minDF,(55.0 to 95.0 by 20.0))
      .addGrid(lr.regParam, Array(10e-8, 10e-6, 10e-4, 10e-2))
      .build()

    // On utilise le F1 score pour comparer nos différents modèles
    val eval = new MulticlassClassificationEvaluator()
      .setLabelCol("final_status")
      .setPredictionCol("predictions")
      .setMetricName("f1")

    // On utilise 70% des données pour l'entrainement et 30% pour la validation
    val trainValidationSplit = new TrainValidationSplit()
      .setEstimator(pipeline)
      .setEvaluator(eval)
      .setEstimatorParamMaps(paramGrid)
      .setTrainRatio(0.7)

    val model = trainValidationSplit.fit(training)

    // l) On affiche les 5 premières prédictions ...
    val df_WithPredictions = model.transform(test)
    df_WithPredictions.select("features","predictions").show(5)

    // Puis le F1 score
    val evaluator = new MulticlassClassificationEvaluator()
      .setLabelCol("final_status")
      .setPredictionCol("predictions")
      .setMetricName("accuracy")

    val accuracy = evaluator.evaluate(df_WithPredictions)
    println("Il y a  " + (accuracy) * 100 + " % de prédictions justes.")
    println("Il y a  " + (1.0 - accuracy) * 100 + " % d'erreur.")

    // m) On affiche une "pseudo matrice" de confusion
    df_WithPredictions.groupBy("final_status", "predictions").count.show()

    // On sauvegarde notre modèle
    //** model.write.save("Regression_Logistique") **//
  }
}
